gemini/use-cases/retrieval-augmented-generation/qna_using_query_routing/utils/intent_routing.py

```diff
@@ -0,0 +1,413 @@
+        4. Strictly answer the question if only {enabled_programming_language} is mentioned in question.
+
+        If the question is about other programming language then DO NOT provide any answer, just say "{non_programming_question_error_msg}"
+
+        """
+        )
+
+        response = chat.send_message(
+            f"""{text}""", generation_config=self.genai_chat_parameters
+        )
+        if response.to_dict()["candidates"][0]["finish_reason"] != 1:
+            self.logger.info(
+                "ask_codey: No response from QnA due to LLM safety checks."
+            )
+            self.logger.info("LLM error code: %s\n", response.raw_prediction_response)
+        response = response.text
+        response = response.replace("```", "\n\n```")
+        response = response.replace("```java", "``` java")
+        return response
+
+    def get_programming_lanuage_from_query(
+        self, text: str, enabled_programming_language: List
+    ) -> List[str]:
+        """
+        Extracts programming languages mentioned in a user's query.
+
+        Args:
+            text (str): The user's query.
+            enabled_programming_language (List): List of supported languages.
+
+        Returns:
+            list: A list of programming languages extracted from the query (potentially empty).
+        """
+
+        response = self.model.generate_content(
+            f"""
+            You are strict programming languages extractor.
+
+            Instructions:
+            1. Extract only programming languages from message.
+            2. Don't return any other languages other than programming.
+            3. return [] if no programming lanuage in mentioned in message.
+            4. {enabled_programming_language} these are the programming languages.
+
+            Examples:
+            write a code for fibonacci series in C++? : ["C++"]
+            write a code using C# to generate palindrome series : ["C#"]
+            using python, write a sample application code to create endpoint? : ["Python"]
+            what are classes in Java? : ["Java"]
+            write a code for reverse a string : []
+            what are data types? : []
+
+            What are the programming languages mentioned in below message?
+            MESSAGE:{text}
+            programming languages:""",  # pylint: disable=C0301:line-too-long
+            generation_config=self.genai_qna_parameters,
+        )
+
+        programming_lang = response.text
+        program_lang_in_query = []
+        if programming_lang:
+            try:
+                programming_lang = programming_lang.replace("'", '"')
+                programming_lang = json.loads(programming_lang)
+                program_lang_in_query = [
+                    x.lower().replace(" ", "").strip() for x in programming_lang
+                ]
+            except Exception:  # pylint: disable=W0718,W0703
+                self.logger.info("Error while extracting programming language.")
```

Comment:
You can use JSON output mode for Gemini 1.5 Pro to avoid these parsing issues. https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/gemini#:~:text=in%20the%20list.-,responseMimeType,-(Preview)